pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -1471,9 +1554,11 @@ def put_object_as_directory(self, ctx):
             rpc_resp)
 
         resp_headers = {
-            "Etag": construct_etag(ctx.account_name, inode, num_writes),
             "Content-Type": DIRECTORY_CONTENT_TYPE,
             "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
+        resp_headers["Etag"] = unmung_etags_and_return_best(
```

Comment:
the short answer is "i don't know."  the longer answer is "i don't know and i didn't really want to spend the time to figure it out."

from looking at the other code, i suspect that the PUT request that creates a directory could be coming from the `s3api` filter and it could supply its own value for the ETag, with the header name `S3API_ETAG_HEADER`.  there could also be a `LISTING_ETAG_OVERRIDE_HEADER`.

so, while a header named `ETag` (or `etag` or `Etag`) should have the value `EMPTY_OBJECT_ETAG`, there could also be a header named `S3API_ETAG_HEADER`, and if that header exists we should(?) return that ETag.

but, looking at the code for `unmung_etags_and_return_best()`, i can see that for a directory that function will only ever return `EMPTY_OBJECT_ETAG`.

i'm not sure if that's the correct behavior or not; i think i just copied what the earlier code seemed to do (see the above comment about "... and i don't really want to spend the time to figure it out.").

frankly, if the s3api had set a value for `LISTING_ETAG_OVERRIDE_HEADER` or `S3API_ETAG_HEADER` for the directory then i think `unmung_etags_and_return_best()` should return that ETag for the directory.

what do you think?

Reply:
i still think the correct behavior here is kind've murky.

_arguably_ if `S3API_ETAG_HEADER` was supplied with the request then that's what should be returned.  however, it is also true that the old `best_possible_etag()` code will return `EMPTY_OBJECT_ETAG` for all directories (or, more accurately, `EMPTY_OBJECT_ETAG` if the _optional_ `is_dir` is true; it default to false).

the original code seemed to have:
```
         resp_headers = {
-            "Etag": construct_etag(ctx.account_name, inode, num_writes),
             "Content-Type": DIRECTORY_CONTENT_TYPE,
             "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
```
which definitely seems wrong.

looking through the code, `get_object()` does not set the optional `is_dir` argument when calling `best_possible_etag()` and `get_object()` does not return the `IsDir` metadta..  however, the current behavior of proxyfsd is for a `get_object()` on a directory to fail with a `500` series error which seems quite wrong.

i have a fix for that (to fix problems that Burton is seeing) but it looks like that fix will need to percolate through the Python `get_object()` code.

anyway, in your opinion, should the ETag for a directory always be `EMPTY_OBJECT_ETAG` regardless of any headers set by the S3API layer?

Reply:
i changed the code again on the theory that `best_possible_etag()` might change in the future.  so now it looks like:
```
        # currently best_possible_etag() returns EMPTY_OBJECT_ETAG for                                                   
        # all directories, but that might change in the future.                                                          
        # unmung the obj_metadata so best_possible_etag() can use it                                                     
        # if its behavior changes (note that num_writes is forced to 0).                                                 
        unmung_etags(obj_metadata, 0)
        resp_headers = {
            "Content-Type": DIRECTORY_CONTENT_TYPE,
            "Last-Modified": last_modified_from_epoch_ns(mtime_ns)}
        resp_headers["ETag"] = best_possible_etag(
            obj_metadata, ctx.account_name, inode, 0, is_dir=True)
```